File: restapi-with-asp.net-core-7-and-inmemory-store/Movies.Api/Store/InMemory/InMemoryMoviesStore.cs
namespace Movies.Api.Store.InMemory;

public class InMemoryMoviesStore : IMoviesStore
{
    private readonly Dictionary<Guid, Movie> repository = new Dictionary<Guid, Movie>();

    public IEnumerable<Movie> GetAll()
    {
        return repository.Values.AsEnumerable();
    }

    public Movie? GetById(Guid id)
    {
        if (repository.ContainsKey(id))
        {
            return repository[id];
        }

        return null;
    }

    public void Create(CreateMovieParams createMovieParams)
    {
        if (repository.ContainsKey(createMovieParams.Id))
        {
            throw new DuplicateKeyException($"Duplicate movie id: {createMovieParams.Id}");
        }

        var movie = new Movie(
            createMovieParams.Id,
            createMovieParams.Title,
            createMovieParams.Director,
            createMovieParams.TicketPrice,
            createMovieParams.ReleaseDate,
            DateTime.UtcNow,
            DateTime.UtcNow);
        repository.Add(movie.Id, movie);
    }

    public void Update(Guid id, UpdateMovieParams updateMovieParams)
    {
        if (!repository.ContainsKey(id))
        {
            throw new RecordNotFoundException();
        }

        var movieToUpdate = repository[id];
        var movie = new Movie(
            movieToUpdate.Id,
            updateMovieParams.Title,
            updateMovieParams.Director,
            updateMovieParams.TicketPrice,
            updateMovieParams.ReleaseDate,
            movieToUpdate.CreatedAt,
            DateTime.UtcNow);

        repository[id] = movie;
    }

    public void Delete(Guid id)
    {
        if (!repository.ContainsKey(id))
        {
            throw new RecordNotFoundException();
        }

        repository.Remove(id);
    }
}

File: restapi-with-go-chi-and-inmemory-store/store/in_memory/in_memory_movies_store.go
package in_memory

import (
	"context"
	"movies-api/store"
	"sync"
	"time"

	"github.com/google/uuid"
)

type InMemoryMoviesStore struct {
	movies map[uuid.UUID]*store.Movie
	mu     sync.RWMutex
}

func NewInMemoryMoviesStore() *InMemoryMoviesStore {
	return &InMemoryMoviesStore{
		movies: map[uuid.UUID]*store.Movie{},
	}
}

func (s *InMemoryMoviesStore) GetAll(ctx context.Context) ([]*store.Movie, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	movies := make([]*store.Movie, 0)
	for _, m := range s.movies {
		movies = append(movies, m)
	}
	return movies, nil
}

func (s *InMemoryMoviesStore) GetById(ctx context.Context, id uuid.UUID) (*store.Movie, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	m, ok := s.movies[id]
	if !ok {
		return nil, &store.RecordNotFoundError{}
	}

	return m, nil
}

func (s *InMemoryMoviesStore) Create(ctx context.Context, createMovieParams store.CreateMovieParams) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, ok := s.movies[createMovieParams.Id]; ok {
		return &store.DuplicateIdError{Id: createMovieParams.Id}
	}

	movie := &store.Movie{
		Id:          createMovieParams.Id,
		Title:       createMovieParams.Title,
		Director:    createMovieParams.Director,
		ReleaseDate: createMovieParams.ReleaseDate,
		TicketPrice: createMovieParams.TicketPrice,
		CreatedAt:   time.Now().UTC(),
		UpdatedAt:   time.Now().UTC(),
	}

	s.movies[movie.Id] = movie
	return nil
}

func (s *InMemoryMoviesStore) Update(ctx context.Context, id uuid.UUID, updateMovieParams store.UpdateMovieParams) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	m, ok := s.movies[id]
	if !ok {
		return &store.RecordNotFoundError{}
	}

	m.Title = updateMovieParams.Title
	m.Director = updateMovieParams.Director
	m.ReleaseDate = updateMovieParams.ReleaseDate
	m.TicketPrice = updateMovieParams.TicketPrice
	m.UpdatedAt = time.Now().UTC()

	s.movies[id] = m
	return nil
}

func (s *InMemoryMoviesStore) Delete(ctx context.Context, id uuid.UUID) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	delete(s.movies, id)
	return nil
}
